File: algorithms/java/112_path_sum/PathSum.java
class PathSum {
	public boolean hasPathSum(TreeNode root, int sum) {
		if (root == null) return false;
		return hasPathSum(root, sum, 0);
	}
	private boolean hasPathSum(TreeNode root, int sum, int cur) {
		cur += root.val;
		if (root.left == null && root.right == null) {
			return sum == cur;
		}
		if (root.left != null && hasPathSum(root.left, sum, cur)) {
			return true;
		}
		if (root.right != null && hasPathSum(root.right, sum, cur)) {
			return true;
		}
		return false;
	}
}

File: algorithms/cpp/112_path_sum/PathSum.cpp
class PathSum {
public:
	bool hasPathSum(TreeNode* root, int target_sum) {
        if (root == NULL)
            return false;
        if (root->left == NULL && root->right == NULL) {
            return target_sum - root->val == 0;
        }
        return hasPathSum(root->left, target_sum - root->val) || hasPathSum(root->right, target_sum - root->val);
    }
};
